tensorboard/data/server/run.rs

```diff
@@ -0,0 +1,363 @@
+}
+
+impl Default for RunLoader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::fs::File;
+    use std::io::{BufWriter, Write};
+
+    /// Writes an event to the given writer, in TFRecord form.
+    fn write_event<W: Write>(writer: W, event: &pb::Event) -> std::io::Result<()> {
+        use prost::Message;
+        let mut data = Vec::new();
+        event.encode(&mut data)?;
+        crate::tf_record::TfRecord::from_data(data).write(writer)?;
+        Ok(())
+    }
+
+    /// Writes a TF 1.x scalar event (`simple_value`) to the given writer.
+    fn write_scalar<W: Write>(
+        writer: W,
+        tag: &Tag,
+        step: Step,
+        wt: WallTime,
+        value: f32,
+    ) -> std::io::Result<()> {
+        let event = pb::Event {
+            step: step.0,
+            wall_time: wt.into(),
+            what: Some(pb::event::What::Summary(pb::Summary {
+                value: vec![pb::summary::Value {
+                    tag: tag.0.clone(),
+                    value: Some(pb::summary::value::Value::SimpleValue(value)),
+                    ..Default::default()
+                }],
+                ..Default::default()
+            })),
+            ..Default::default()
+        };
+        write_event(writer, &event)
+    }
+
+    #[test]
+    fn test() -> Result<(), Box<dyn std::error::Error>> {
+        let logdir = tempfile::tempdir()?;
+        let f1_name = logdir.path().join("tfevents.123");
+        let f2_name = logdir.path().join("tfevents.456");
+        let mut f1 = BufWriter::new(File::create(&f1_name)?);
+        let mut f2 = BufWriter::new(File::create(&f2_name)?);
+
+        // Write file versions.
+        for (f, wall_time) in &mut [(&mut f1, 1234.0), (&mut f2, 2345.0)] {
+            let file_version = pb::Event {
+                wall_time: *wall_time,
+                what: Some(pb::event::What::FileVersion("brain.Event:2".to_string())),
+                ..Default::default()
+            };
+            write_event(f, &file_version)?;
+        }
+
+        // Write some data points across both files.
+        let tag = Tag("accuracy".to_string());
+        write_scalar(&mut f1, &tag, Step(0), WallTime::new(1235.0).unwrap(), 0.25)?;
+        write_scalar(&mut f1, &tag, Step(1), WallTime::new(1236.0).unwrap(), 0.50)?;
+        write_scalar(&mut f2, &tag, Step(2), WallTime::new(2346.0).unwrap(), 0.75)?;
+        write_scalar(&mut f2, &tag, Step(3), WallTime::new(2347.0).unwrap(), 1.00)?;
+        f1.into_inner()?.sync_all()?;
+        f2.into_inner()?.sync_all()?;
+
+        let mut loader = RunLoader::new();
+        loader.reload(vec![f1_name, f2_name]);
+
+        // Start time should be that of the file version event, even though that didn't correspond
+        // to any time series.
+        assert_eq!(loader.start_time, Some(WallTime::new(1234.0).unwrap()));
+        assert_eq!(loader.time_series.keys().collect::<Vec<_>>(), vec![&tag]);
```

Comment:
`<_>`! One day I will see a code with my favorite emoji, `>_<`

Reply:
Hmm… I’m having trouble thinking of a piece of standard Rust syntax that
contains `>_<`, since `_` is not a valid identifier by itself. :-(
But this time, [macros *can* save the day][play]:

```rust
macro_rules! swlmoji {
    (<_>) => { "meep" };
    (>_<) => { "mrrp" };
}
fn main() {
    println!("{}", swlmoji!(<_>));
    println!("{}", swlmoji!(>_<));
}
```

[play]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cac3ec76b5687aff4fb8775b19217dc0

Reply:
(oh, and the `_` in `Vec<_>` means “please infer type parameter :-) )